src/txmempool.cpp

```diff
@@ -120,47 +120,91 @@ void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap& cachedDescendan
+                            if (!(descendants.size() == i+2)) {
+                                std::swap(descendants[i+1], descendants.back());
+                            }
+                            // skip processing this element
+                            ++i;
+                        }
+                    }
+                } else if (!visited(mapTx.iterator_to(childEntry))) {
+                    // Schedule for later processing
+                    descendants.emplace_back(mapTx.iterator_to(childEntry));
+                    ++n_to_process;
                 }
-            } else if (!descendants.count(childEntry)) {
-                // Schedule for later processing
-                stageEntries.insert(childEntry);
             }
         }
-    }
+    } // release epoch
+
+    // remove any descendants that are in setExclude
+    auto included_upto = std::remove_if(descendants.begin(), descendants.end(),
+            [&](txiter it) {
+                return setExclude.count(it->GetTx().GetHash());
+            });
+
+    // if none remain, we don't have to do any updating
+    if (included_upto == descendants.begin()) return;
+
     // descendants now contains all in-mempool descendants of updateIt.
     // Update and add to cached descendant map
     int64_t modifySize = 0;
     CAmount modifyFee = 0;
     int64_t modifyCount = 0;
-    for (const CTxMemPoolEntry& descendant : descendants) {
-        if (!setExclude.count(descendant.GetTx().GetHash())) {
-            modifySize += descendant.GetTxSize();
-            modifyFee += descendant.GetModifiedFee();
-            modifyCount++;
-            cachedDescendants[updateIt].insert(mapTx.iterator_to(descendant));
-            // Update ancestor state for each descendant
-            mapTx.modify(mapTx.iterator_to(descendant), update_ancestor_state(updateIt->GetTxSize(), updateIt->GetModifiedFee(), 1, updateIt->GetSigOpCost()));
-            // Don't directly remove the transaction here -- doing so would
-            // invalidate iterators in cachedDescendants. Mark it for removal
-            // by inserting into descendants_to_remove.
-            if (descendant.GetCountWithAncestors() > ancestor_count_limit || descendant.GetSizeWithAncestors() > ancestor_size_limit) {
-                descendants_to_remove.insert(descendant.GetTx().GetHash());
-            }
+    // Note: the below contains code which does some hacks to keep memory tight.
+    // it could be improved in the future to detect if the vector is already tight
+    // and then directly move it to cachedDescendants. For simplicity, we just
+    // do a copy for now.
+
+    // emplace into a new vector to guarantee we trim memory
+    const auto& it = cachedDescendants.emplace(std::piecewise_construct,
+            std::forward_as_tuple(updateIt),
+            std::forward_as_tuple(descendants.begin(), included_upto));
+    // swap with descendants to release it early!
+    std::vector<txiter>().swap(descendants);
```

Comment:
Also I'm curious as to why it's necessary to free this memory here. We don't need to allocate anything in the rest of the function, and `descendants` goes out of scope when we return.

Reply:
from what you cited:
>The request is non-binding, and the container implementation is free to optimize otherwise and leave the [vector](https://www.cplusplus.com/vector) with a [capacity](https://www.cplusplus.com/vector::capacity) greater than its [size](https://www.cplusplus.com/vector::size).

Ideally we would be happy trusting our shrink_to_fit, but I'd rather get exactly a trimmed vector here.

we do allocate (descendants_to_remove.insert), so it's a cautious approach to free it explicitly when it will never be used again than later.

Reply:
Sounds good.

And my other question - what's the point of releasing `descendants` early?

Reply:
> e do allocate (descendants_to_remove.insert), so it's a cautious approach to free it explicitly when it will never be used again than later.